src/main/scala/org/broadinstitute/hail/driver/GRM.scala

```diff
@@ -72,32 +75,70 @@ object GRM extends Command {
       && options.nFile != null)
       warn(s"format ${options.format}: ignoring `--N-file'")
 
+    val tmpFileName = state.hadoopConf.getTemporaryFile(HailConfiguration.tmpDir)
+
     options.format match {
       case "rel" =>
-        state.hadoopConf.writeTextFile(options.output) { s =>
-          for (i <- 0 until nSamples) {
-            for (j <- 0 to i) {
+        grm.toIndexedRowMatrix()
+          .rows
+          .sortBy(row => row.index)
+          .map { (row: IndexedRow) =>
+            val i = row.index
+            val arr = row.vector.toArray
+            val sb = new StringBuilder
+            var j = 0
+            while (j <= i) {
               if (j > 0)
-                s.write("\t")
-              s.write(grm(i, j).toString)
+                sb.append("\t")
+              sb.append(arr(j))
+              j += 1
             }
-            s.write("\n")
+            sb.toString()
           }
-        }
+          .saveAsTextFile(tmpFileName)
+
+        state.hadoopConf.copyMerge(tmpFileName, options.output, true, false)
 
       case "gcta-grm" =>
-        state.hadoopConf.writeTextFile(options.output) { s =>
-          for (i <- 0 until nSamples)
-            for (j <- 0 to i)
-              s.write(s"${i + 1}\t${j + 1}\t$nVariants\t${grm(i, j)}\n")
-        }
+        grm.toIndexedRowMatrix()
+          .rows
+          .sortBy(row => row.index)
+          .map { (row: IndexedRow) =>
+            val i = row.index
+            val arr = row.vector.toArray
+            val sb = new StringBuilder
+            var j = 0
+            while (j <= i) {
+              sb.append(s"${i + 1}\t${j + 1}\t$nVariants\t${arr(j)}")
```

Comment:
I did a few tests and whatever difference there is between these two approaches is in the noise of computing the GRM. I think we should stick with the string interpolation because, IMO, it's much easier to read and manipulate.